lib/qbo_api/connection/oauth2.rb

```diff
@@ -1,5 +1,49 @@
 class QboApi
-  attr_accessor :access_token
+  attr_accessor :client_id, :client_secret, :refresh_proc
+
+  OAUTH2_BASE           = 'https://oauth.platform.intuit.com'
+  OAUTH2_URL            =  OAUTH2_BASE + '/oauth2/v1/tokens/bearer'
+
+  def access_token
+    @access_token || (refresh_access_token! if refresh_token)
+  end
+
+  def access_token=(token)
+    reset_connections
+    @access_token = token
+  end
+
+  def refresh_token
+    @refresh_token
+  end
+
+  def refresh_token=(token)
+    reset_connections if refresh_token
+    refresh_proc.call(token) if refresh_proc and refresh_token
+    @refresh_token = token
+  end
+
+  def refresh_access_token!
+    resp = oauth2_connection.post do |req|
+      req.body = {
+          grant_type: :refresh_token,
+          refresh_token: refresh_token
+      }
+    end
+    data = parse_response_body(resp)
+    self.refresh_token = data.fetch('refresh_token')
+    self.access_token = data.fetch('access_token')
+  end
+
+  def oauth2_connection
+    headers = {'Accept' => 'application/json'}
+    @oauth2_connection ||= build_connection(OAUTH2_URL, headers: headers) do |conn|
```

Comment:
this isn't really the right method name or variable names

the url I call the `token_endpoint`

also, this should be re-used in the example apps imho

my own code looks like this:

```ruby
    def refresh_token
      cache.fetch("quickbooks.refresh_token")
    end

    def refresh_token=(refresh_token)
      fail ArgumentError, "not allowed to set nil refresh_token" if refresh_token.nil?
      if Integer(refresh_token_expires_in) > 120
        Rails.logger.info "[Quickbooks] Changing refresh token even though expiry=#{refresh_token_expires_in}"
      end
      cache.write("quickbooks.refresh_token", refresh_token, expires_in: 31536000) # 1 year == 60 * 60 * 24 * 365
    end

    def refresh_token_clear
      cache.delete("quickbooks.refresh_token")
    end

    def access_token
      cache.fetch("quickbooks.access_token") do
        access_token!
      end
    end

    def access_token!
      json = refresh_access_token
      json.fetch(:access_token) or fail ArgumentError, "not allowed to set nil access_token"
    end

    def access_token=(access_token)
      fail ArgumentError, "not allowed to set nil access_token" if access_token.nil?
      if Integer(access_token_expires_in) > 10
        Rails.logger.info "[Quickbooks] Changing access token even though expiry=#{access_token_expires_in} refresh_token_expires_in=#{refresh_token_expires_in}"
      end
      cache.write("quickbooks.access_token", access_token, expires_in: 31536000) # 1 year == 60 * 60 * 24 * 365
    end

    def access_token_clear
      cache.delete("quickbooks.access_token")
    end

    # e.g. 8706412
    def refresh_token_expires_in
      cache.read("quickbooks.refresh_token_expires_in") || -127
    end

    # e.g. 3600
    def access_token_expires_in
      cache.read("quickbooks.access_token_expires_in") || -127
    end

    def refresh_access_token(current_refresh_token: refresh_token)
      fail ArgumentError, "refresh_token cannot be nil" if current_refresh_token.nil?
      url = 'https://oauth.platform.intuit.com'
      token_endpoint = '/oauth2/v1/tokens/bearer'

      connection = Faraday.new(url: url) { |conn|
        conn.basic_auth(client_id, client_secret)
        conn.request :url_encoded # application/x-www-form-urlencoded
        conn.headers['Accept'] = "application/json"

        conn.use FaradayMiddleware::ParseJson, parser_options: { symbolize_names: true }
        conn.use Faraday::Response::RaiseError
        conn.response :detailed_logger, Auth.logger if QboApi.log
        conn.adapter  Faraday.default_adapter
      }
      synchronize(__callee__) do
        raw_response = connection.post {|req|
          req.body = { grant_type: :refresh_token, refresh_token: current_refresh_token }
          req.url token_endpoint
        }

        data = raw_response.body
        # {
        #   "expires_in": 3600,
        #   "x_refresh_token_expires_in": 8706412,
        #   "refresh_token": "newnew...",
        #   "access_token": "new...",
        #   "token_type": "bearer"
        # }
        new_refresh_token = data.fetch(:refresh_token)
        if new_refresh_token != current_refresh_token
          Auth.logger.info { "Quickbooks RefreshToken changed to '#{new_refresh_token}' from '#{current_refresh_token}'" }
        end
        if new_refresh_token.nil? || new_refresh_token.empty?
          Auth.logger.info { "Quickbooks returned an empty refresh token: #{data.inspect}" }
          fail ArgumentError, "Quickbooks returned an empty refresh token: #{data.inspect}"
        end
        cache.write("quickbooks.refresh_token_expires_in", data.fetch(:x_refresh_token_expires_in))
        cache.write("quickbooks.access_token_expires_in", data.fetch(:expires_in))
        self.refresh_token = new_refresh_token
        self.access_token = data.fetch(:access_token)
        data
      end
    rescue Faraday::ClientError => e
      data ||= :nil_data
      status = defined?(raw_response) ? raw_response.status : :no_response
      ExceptionNotifier.error(e, "[Quickbooks] Failed to refresh token", response: { data: data.inspect, status: status }, client_error: e.inspect)
      raise
    end

    # Relevant Quickbooks docs: Get authorization code
    # - https://developer.intuit.com/docs/0100_quickbooks_online/0100_essentials/000500_authentication_and_authorization/connect_from_within_your_app#Refreshing_the_access_token
    # Relevant Quickbooks docs: Exchange authorization code for refresh and access tokens
    # - https://developer.intuit.com/docs/0100_quickbooks_online/0100_essentials/000500_authentication_and_authorization/connect_from_within_your_app#Exchange_code_for_refresh_and_access_tokens
    def authorize
      url = 'https://appcenter.intuit.com'
      authorization_endpoint = '/connect/oauth2'
      csrf_token = SecureRandom.uuid # should be stored, used in redirect handling
      cache.write("quickbooks.csrf_token", csrf_token)

      connection = Faraday.new(url: url) { |conn|
        conn.use Faraday::Response::RaiseError
        conn.response :detailed_logger, Auth.logger if QboApi.log
        conn.adapter  Faraday.default_adapter
      }
      raw_response = connection.get {|req|
        params = {
          client_id: Auth.client_id,
          scope: "com.intuit.quickbooks.accounting",
          response_type: "code",
          state: csrf_token,
          redirect_uri: Auth.oauth2_redirect_uri
        }
        req.url authorization_endpoint, params
      }
      # The response is sent to the redirect_uri that you specified in the request. All responses are returned in the query string, as shown below:
      raw_response
      # raw_response.env[:url]
      # #=> https://appcenter.intuit.com/connect/oauth2?client_id=<CLIENT_ID>&redirect_uri=http%3A%2F%2Flocalhost%3A5000%2Fquickbooks%2Foauth2_callback&response_type=code&scope=com.intuit.quickbooks.accounting&state=<CSRF_TOKEN>
    end

    # Authorization request redirect handling:
    #  - uri structure: #{redirect_uri}?state=#{csrf_token}%26url%3D#{redirect_uri}&code=#{authentication_code}&realmId=#{realm_id}
    #  - `code`: the authorization code, used in +exchange_authorization_code_for_access_token+
    #  - `realmId`: used by when authenticating API client
    #  - `state`: confirm matches the `csrf_token` declared above.
    #  - `error`: one of access_denied or invalid_scope
    # @example:
    #   GET "/quickbooks/oauth2_callback?state=<CSRF_TOKEN>&code=<AUTH_CODE>&realmId=<REALM_ID>
    def handle_authorization_redirect(params)
      error = params[:error]
      raise "Authorization Error: #{error.inspect}" if error
      csrf_token = cache.read("quickbooks.csrf_token")
      cache.delete("quickbooks.csrf_token")
      raise "CSRF token did not match" if csrf_token != params.fetch(:state)
      raise "Unexpected realmid #{params.fetch(:realmId)}" if Auth.realm_id != params.fetch(:realmId)
      authorization_code = params.fetch(:code)
      exchange_authorization_code_for_access_token(authorization_code)
    end

    def exchange_authorization_code_for_access_token(authorization_code)
      url = 'https://oauth.platform.intuit.com'
      token_endpoint = '/oauth2/v1/tokens/bearer'

      connection = Faraday.new(url: url) { |conn|
        conn.basic_auth(client_id, client_secret)
        conn.request :url_encoded # application/x-www-form-urlencoded
        conn.headers['Accept'] = "application/json"

        conn.use FaradayMiddleware::ParseJson, parser_options: { symbolize_names: true }
        conn.use Faraday::Response::RaiseError
        conn.response :detailed_logger, Auth.logger if QboApi.log
        conn.adapter  Faraday.default_adapter
      }
      raw_response = connection.post {|req|
        req.body = { code: authorization_code, grant_type: :authorization_code, redirect_uri: Auth.oauth2_redirect_uri }
        req.url token_endpoint
      }
      data = raw_response.body
      # {
      #   "expires_in": 3600,
      #   "x_refresh_token_expires_in": 8706412,
      #   "refresh_token": "newnew...",
      #   "access_token": "new...",
      #   "token_type": "bearer"
      # }
      yield data if block_given?
      new_refresh_token = data.fetch(:refresh_token)
      self.refresh_token = new_refresh_token
      self.access_token = data.fetch(:access_token)
      data
    end

    def self.client(current_access_token: access_token, realm_id: realm_id())
      raise ArgumentError, "access_token cannot be nil" if current_access_token.nil?
      raise ArgumentError, "realm_id cannot be nil" if realm_id.nil?
      client = QboApi.new(access_token: current_access_token, realm_id: realm_id)
      Client.new(client) # my wrapper for QboApi
    end

    def self.refresh_client(client)
      refresh_access_token
      client(realm_id: client.realm_id)
    end

    def self.synchronize(lock_name, expires_in_seconds: 2, lock_is_exclusive_until_expires: 0)
      caller_location = caller_locations.first
      options = {
        trace: { file: caller_location.path, line: caller_location.lineno, method: caller_location.label },
        pid: Process.pid, sidekiq_context: Thread.current[:sidekiq_context]
      }.compact
      mutex = RedisMutex.new(lock_name, block: lock_is_exclusive_until_expires, expire: expires_in_seconds)
      if mutex.lock
        Rails.logger.debug do "Mutex: running in lock: options=#{options.inspect}" end
        yield.tap do
          Rails.logger.debug do "Mutex: completed in lock: options=#{options.inspect}" end
          next if mutex.unlock
          Rails.logger.warn do "Mutex: failed to unlock: options=#{options.inspect}" end
          ExceptionNotifier.log(:warn, "Mutex: failed to unlock", *options)
        end
      else
        Rails.logger.debug do "Mutex: failed to acquire lock: options=#{options.inspect}" end
      end
    end

```

my client error handling code includes the execution of the refresh

```ruby
      # Expected Client errors
      #   qbo_api exceptions https://github.com/minimul/qbo_api/blob/master/lib/qbo_api/raise_http_exception.rb
      #   e.fault[:error_body].first[:fault_type] # =>"ValidationFault"
      #   e.fault[:error_body].first[:error_code]
      #     "6000" |
      #            | uuid already used: "An unexpected error occurred while accessing or saving your data. Please wait a few minutes and try again.  If the problem persists, contact customer support."
      #            | need a mutex: "You can only add or edit one name at a time.  Please try again."
      #            |   In the application layer there is a lock across Employee, Customer and Vendor entities to ensure unique name constraint. The above error comes, when you try to modify/create multiple objects using parallel threads. Please call API sequentially for the above 3 name entities. If you still face any error, you can update your post with corresponding request(and API requestID) and response XML.
      #            |   This is a valid error caused by contention. Please use Batch operation instead for sending multiple requests. https://developer.intuit.com/docs/0100_quickbooks_online/0300_references/0000_programming_guide/0000_rest_api_quick_reference#/Batch_operations
      #     "6240" | name not unique
      #     "4000" | Query Error, see e.fault[:error_detail]
      #      "610" | entity not found
      #   e.fault[:error_body].first[:fault_type] # =>"AUTHENTICATION"
      #   e.fault[:error_body].first[:error_code]
      #     "3200" | unauthorized
      #   faraday http client exceptions https://github.com/lostisland/faraday/blob/master/lib/faraday/response/raise_error.rb
      def handle_errors(&block)
        result = ::Result.new(&block)
        return result if result.ok?
        case result.error
        when QboApi::Unauthorized
          Auth.logger.info "Quickbooks access_token has expired. Refreshing and trying again"
          @quickbooks_client = Auth.refresh_client(quickbooks_client).quickbooks_client
          ::Result.new(&block).tap do |new_result|
            if new_result.ok?
              Auth.logger.info "Quickbooks access_token has been refreshed and request retry succeeded: #{new_result}"
            else
              notify_retry_failed(result, new_result)
            end
          end
        else
          result
        end
      end
```

The Result object is basically github.com/bf4/safety_dance

and in my routes

```ruby
  post ENV.fetch("QUICKBOOKS_WEBHOOK_PATH") => ->(env) do
    request = Rack::Request.new(env)
    env['rack.input'].rewind
    body = env['rack.input'].read
    hmac_header = env['HTTP_INTUIT_SIGNATURE']
    digest  = OpenSSL::Digest.new('sha256')
    verifier_token = ENV["QUICKBOOKS_WEBHOOK_VERIFIER_TOKEN"]
    calculated_hmac = Base64.encode64(OpenSSL::HMAC.digest(digest, verifier_token, body)).strip
    if calculated_hmac == hmac_header
      Rails.logger.info "[WEBHOOK:QUICKBOOKS] Verified: #{body}"
    else
      Rails.logger.warn "[WEBHOOK:QUICKBOOKS] Not Verified: #{body} hmac_header=#{hmac_header} calculated_hmac=#{calculated_hmac} verifier_token_set=#{verifier_token.present?}"
    end
    [201, {}, []]
  end

  get ENV.fetch('QUICKBOOKS_OAUTH2_AUTHORIZATION_PATH') => ->(env) {
    data = Accounting::Auth.authorize
    [302, {'Location' => data.env[:url].to_s}, []]
  }
  get URI(Accounting::Auth.oauth2_redirect_uri).path => ->(env) {
    request = Rack::Request.new(env)
    params = ActionController::Parameters.new(request.params)
      .permit(:code, :realmId, :state, :error)
    data = Accounting::Auth.handle_authorization_redirect(params)
    body = <<-BODY
<!DOCTYPE html> <html lang="en"> <head> <meta charset="UTF-8"> <title></title> <style type="text/css" media="all"> ul li span { vertical-align: top; } </style> </head>
<body> <h1>Account authorized</h1> <pre><span>Accounting::Auth.access_token =</span> "<span>#{data.fetch(:access_token)}</span>"<br><span>Accounting::Auth.refresh_token =</span> "<span>#{data.fetch(:refresh_token)}</span>"</pre></body>
</html>
    BODY
    [200, {}, [body]]
  }
```